### PR TITLE
[Bugfix] Handle numpy array outputs when generate image

### PR DESCRIPTION
**Pull request by lengrongfu**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose

Fixes: https://github.com/vllm-project/vllm-omni/pull/1660

Wan2.2 video pipelines return numpy arrays (output_type="np"), but image_api_utils.py assumes PIL Images and calls img.save(). 

We `_extract_images_from_result` this method to add a numpy-to-PIL conversion.

Serve:
```
vllm serve /home/jovyan/Wan2.2-T2V-A14B-Diffusers/ --omni --port 8091
```

Test:
```
curl  -X POST "http://localhost:8091/v1/images/generations"   -H "Content-Type: application/json"   --data '{
            "model": "/home/jovyan/Wan2.2-T2V-A14B-Diffusers/",
            "prompt": "a cat",
            "n": 1,
            "size": "256x256",
            "seed": 7
}' | jq -r '.data[0].b64_json' | base64 -d > cat.png
```

cat

```
curl -X POST "http://localhost:8000/v1/images/generations" \
  -H "Content-Type: application/json" \
  --data '{
    "model": "/home/jovyan/Wan2.2-T2V-A14B-Diffusers/",
    "prompt": "a cat",
    "n": 2,
    "size": "256x256",
    "seed": 7
  }' \
| jq -r '.data[].b64_json' \
| nl -v 0 \
| while read i img; do
    echo "$img" | base64 -d > "cat${i}.png"
  done
```

cat0
cat1

## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by hsliuustc0106** [comment]
why use /v1/images/generations as the endpoint?

**Comment by lengrongfu** [comment]
Sorry, I checked the example and it shows that this API is used. Is it not usable?

File: vllm_omni/entrypoints/openai/api_server.py [hunk 1]
```diff
@@ -1407,7 +1421,9 @@ def _extract_images_from_result(result: Any) -> list[Any]:
             images = request_output["images"]
         elif hasattr(request_output, "images") and request_output.images:
             images = request_output.images
-    return images
+    if isinstance(images[0], np.ndarray):
```
Comment (hsliuustc0106) [inline vllm_omni/entrypoints/openai/api_server.py]: This will raise IndexError if `images` is empty. Add a guard: `if images and isinstance(images[0], np.ndarray):`

File: vllm_omni/entrypoints/openai/api_server.py [hunk 2]
```diff
@@ -1407,7 +1421,9 @@ def _extract_images_from_result(result: Any) -> list[Any]:
...
+        images = images[0]
```
Comment (hsliuustc0106) [inline vllm_omni/entrypoints/openai/api_server.py]: This logic is confusing. If `images[0]` is a numpy array, you replace the entire `images` list with `images[0]`. Is this handling a nested array structure like `[np_array_of_frames]`? Add a comment explaining the expected input format.

File: vllm_omni/entrypoints/openai/api_server.py [hunk 3]
```diff
@@ -1397,6 +1398,19 @@ def _update_if_not_none(object: any, key: str, val: any) -> None:
         setattr(object, key, val)
 
 
+def _normalize_image(image: Any) -> Any:
```
Comment (hsliuustc0106) [inline vllm_omni/entrypoints/openai/api_server.py:1675]: The function handles floating dtypes but doesn't check for unsupported types (e.g., bool, complex). Consider adding: `if not np.issubdtype(image.dtype, np.integer) and not np.issubdtype(image.dtype, np.floating): raise ValueError(f"Unsupported dtype: {image.dtype}")`

File: vllm_omni/entrypoints/openai/api_server.py [hunk 4]
```diff
@@ -1397,6 +1398,19 @@ def _update_if_not_none(object: any, key: str, val: any) -> None:
...
+    """Normalize a single image output to a PIL-compatible format."""
+    if isinstance(image, np.ndarray):
+        while image.ndim > 3:
```
Comment (hsliuustc0106) [inline vllm_omni/entrypoints/openai/api_server.py:1684]: The normalization assumes float range is either [-1, 1] or [0, 1]. What about other ranges? Consider adding a debug log or warning when rescaling from [-1, 1] range to help diagnose issues.

File: vllm_omni/entrypoints/openai/api_server.py [hunk 5]
```diff
@@ -1407,7 +1421,9 @@ def _extract_images_from_result(result: Any) -> list[Any]:
             images = request_output["images"]
         elif hasattr(request_output, "images") and request_output.images:
             images = request_output.images
-    return images
+    if isinstance(images[0], np.ndarray):
+        images = images[0]
+    return [_normalize_image(img) for img in images]
```
Comment (hsliuustc0106) [inline vllm_omni/entrypoints/openai/api_server.py:1726]: Missing unit test for `_normalize_image` and `_extract_images_from_result`. At minimum, add tests for: (1) numpy array with float [0,1], (2) numpy array with float [-1,1], (3) numpy array with int, (4) already PIL Image, (5) empty list.

**Review by hsliuustc0106 — CHANGES_REQUESTED** [review]
## Review

**Rating**: 6/10 | **Verdict**: ⚠️ Changes Requested

### Summary
Bugfix addresses valid issue (Wan2.2 returning numpy arrays), but implementation has critical edge cases and missing tests. Code logic is unclear and needs refactoring.

### Issues

1. **Critical: IndexError on empty list** - Line 1424 checks `images[0]` without verifying the list is non-empty. Will crash if `images` is empty.

2. **Confusing logic** - Line 1425 `images = images[0]` is unclear. Is this handling a nested array case? Needs comment explaining the expected structure.

3. **Missing tests** - No unit tests for the new `_normalize_image` function or the modified `_extract_images_from_result`.

4. **Incomplete PR description** - Checklist items not filled, no test results provided.

5. **Edge cases not handled**:
   - Empty images list
   - Mixed types (some PIL, some numpy)
   - Unsupported dtype (e.g., bool, complex)

### Highlights
- ✅ Addresses real issue (Wan2.2 compatibility)
- ✅ Handles multiple normalization cases (float [0,1] and [-1,1])
- ✅ Properly handles ndim > 3 case

### Recommendation
Fix edge cases and add tests before merge.

---

Reviewed by OpenClaw with [vllm-omni-skills](https://github.com/hsliuustc0106/vllm-omni-skills) 🦐

File: vllm_omni/entrypoints/openai/api_server.py (hunk 2 above)
Comment (lengrongfu) [reply vllm_omni/entrypoints/openai/api_server.py]: yes, when generate more one image this array structure is np_array_of_frames. like:
```
images = np.array([
    [[...], [...], ...],  # one png (256, 256, 3)
    [[...], [...], ...],  # two png (256, 256, 3)
    [[...], [...], ...],  # three png (256, 256, 3)
])
# shape: (3, 256, 256, 3)
```

File: vllm_omni/entrypoints/openai/api_server.py (hunk 1 above)
Comment (lengrongfu) [reply vllm_omni/entrypoints/openai/api_server.py]: thanks tips ~

File: vllm_omni/entrypoints/openai/api_server.py (hunk 3 above)
Comment (lengrongfu) [reply vllm_omni/entrypoints/openai/api_server.py:1675]: thanks tips ~

File: vllm_omni/entrypoints/openai/api_server.py (hunk 4 above)
Comment (lengrongfu) [reply vllm_omni/entrypoints/openai/api_server.py:1684]: add warning log.

File: vllm_omni/entrypoints/openai/api_server.py (hunk 5 above)
Comment (lengrongfu) [reply vllm_omni/entrypoints/openai/api_server.py:1726]: add unit test

**Comment by gcanlin** [comment]
@lishunyang12 PTAL. Thanks!

**Comment by lishunyang12** [comment]
Will take a look.

File: vllm_omni/entrypoints/openai/api_server.py [hunk 6]
```diff
@@ -1397,6 +1398,30 @@ def _update_if_not_none(object: any, key: str, val: any) -> None:
         setattr(object, key, val)
 
 
+def _normalize_image(image: Any) -> Any:
```
Comment (lishunyang12) [inline vllm_omni/entrypoints/openai/api_server.py:1675]: The dtype check runs before the `isinstance(image, np.ndarray)` guard. If `image` is already a PIL Image (which is the common case for non-video pipelines), `image.dtype` raises `AttributeError`. Move the dtype validation inside the `isinstance` block, or add an early return for PIL Images:

```suggestion
def _normalize_image(image: Any) -> Any:
    """Normalize a single image output to a PIL-compatible format."""
    if isinstance(image, Image.Image):
        return image
    if not isinstance(image, np.ndarray):
        raise ValueError(f"Unsupported image type: {type(image)}")
    if not np.issubdtype(image.dtype, np.integer) and not np.issubdtype(image.dtype, np.floating):
        raise ValueError(f"Unsupported dtype: {image.dtype}")
```

File: vllm_omni/entrypoints/openai/api_server.py [hunk 7]
```diff
@@ -1407,7 +1432,11 @@ def _extract_images_from_result(result: Any) -> list[Any]:
             images = request_output["images"]
         elif hasattr(request_output, "images") and request_output.images:
             images = request_output.images
-    return images
+    # Handle when generate more one image
```
Comment (lishunyang12) [inline vllm_omni/entrypoints/openai/api_server.py]: `len(images[0]) > 1` returns the size of the first axis. If a pipeline returns individual `(H, W, C)` arrays in a list, `len` returns `H` (always > 1), so this unwraps a single image into H row-slices. You probably want a shape check instead:

```suggestion
    if images and isinstance(images[0], np.ndarray) and images[0].ndim == 4:
        # Unwrap batch: (N, H, W, C) -> [img1, img2, ...]
        images = list(images[0])
```

**Review by lishunyang12 — DISMISSED** [review]
The idea is right but there are two bugs in the implementation.

**Review by gcanlin — COMMENTED** [review]
Could you upload the generated images?

**Comment by lengrongfu** [comment]
> Could you upload the generated images?

Uploaded, currently, generating multiple photos via HTTP is quite inconvenient, as the response consists of JSON data. I would appreciate it if you could take a look at this PR https://github.com/vllm-project/vllm-omni/pull/1673, which proposes returning the images directly as files. @gcanlin

**Review by hsliuustc0106 — APPROVED** [review]
